File: modules/db/src/blaze/db/impl/index/index_handle.clj
(ns blaze.db.impl.index.index-handle
  (:refer-clojure :exclude [conj])
  (:require
   [blaze.db.impl.codec :as codec])
  (:import
   [blaze.db.impl.index IndexHandle SingleVersionId]))

(set! *warn-on-reflection* true)

(defn from-single-version-id
  "Creates an index handle from `single-version-id`."
  [single-version-id]
  (IndexHandle/fromSingleVersionId ^SingleVersionId single-version-id))

(defn from-resource-handle
  "Creates an index handle from `resource-handle`."
  [resource-handle]
  (IndexHandle/fromIdAndHash (codec/id-byte-string (:id resource-handle))
                             (:hash resource-handle)))

(defn id
  "Returns the id of `index-handle`."
  [index-handle]
  (.id ^IndexHandle index-handle))

(defn hash-prefixes
  "Returns a list of hash prefixes as integers of `index-handle`."
  [index-handle]
  (.hashPrefixes ^IndexHandle index-handle))

(defn to-single-version-ids [index-handle]
  (.toSingleVersionIds ^IndexHandle index-handle))

(defn matches-hash?
  "Tests whether `index-handle` contains the prefix of `hash`."
  [index-handle hash]
  (.matchesHash ^IndexHandle index-handle hash))

(defn conj
  "Returns a new index-handle with the hash of `single-version-id` added to the
  hashes of `index-handle`."
  [index-handle single-version-id]
  (.conj ^IndexHandle index-handle single-version-id))

(defn id-comp
  "Compares `index-handle-1` and `index-handle-2` by id."
  [index-handle-1 index-handle-2]
  (.compareTo (.id ^IndexHandle index-handle-1) (.id ^IndexHandle index-handle-2)))

(defn intersection
  "Returns a new index-handle with the intersection of the hashes found in both
  `index-handle-1` and `index-handle-2`.

  Throws an exception if the IDs of both index handles aren't equal."
  [index-handle-1 index-handle-2]
  (.intersection ^IndexHandle index-handle-1 index-handle-2))

(defn union
  "Returns a new index-handle with the union of the hashes found in both
  `index-handle-1` and `index-handle-2`.

  Throws an exception if the IDs of both index handles aren't equal."
  [index-handle-1 index-handle-2]
  (.union ^IndexHandle index-handle-1 index-handle-2))

File: modules/db/java/blaze/db/impl/index/IndexHandle.java
package blaze.db.impl.index;

import blaze.ByteString;
import blaze.fhir.Hash;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

import static java.util.Objects.requireNonNull;

public final class IndexHandle {

    private final ByteString id;
    private final int[] hashPrefixes;

    private IndexHandle(ByteString id, int[] hashPrefixes) {
        this.id = requireNonNull(id);
        this.hashPrefixes = requireNonNull(hashPrefixes);
    }

    public static IndexHandle fromIdAndHash(ByteString id, Hash hash) {
        return new IndexHandle(id, new int[]{hash.prefix()});
    }

    public static IndexHandle fromSingleVersionId(SingleVersionId singleVersionId) {
        return new IndexHandle(singleVersionId.id(), new int[]{singleVersionId.hashPrefix()});
    }

    public ByteString id() {
        return id;
    }

    public List<Integer> hashPrefixes() {
        return Arrays.stream(hashPrefixes).boxed().toList();
    }

    public List<SingleVersionId> toSingleVersionIds() {
        return Arrays.stream(hashPrefixes).mapToObj(p -> new SingleVersionId(id, p)).toList();
    }

    public boolean matchesHash(Hash hash) {
        return Arrays.binarySearch(hashPrefixes, hash.prefix()) >= 0;
    }

    public IndexHandle conj(SingleVersionId singleVersionId) {
        int hashPrefix = singleVersionId.hashPrefix();
        int index = Arrays.binarySearch(hashPrefixes, hashPrefix);
        if (index >= 0) {
            return this;
        } else {
            index = -(index + 1);
            var newHashPrefixes = new int[hashPrefixes.length + 1];
            System.arraycopy(hashPrefixes, 0, newHashPrefixes, 0, hashPrefixes.length);
            System.arraycopy(hashPrefixes, index, newHashPrefixes, index + 1, hashPrefixes.length - index);
            newHashPrefixes[index] = hashPrefix;
            return new IndexHandle(id, newHashPrefixes);
        }
    }

    public IndexHandle intersection(IndexHandle other) {
        checkIds(other);

        int i = 0, j = 0, r = 0;
        var res = new int[Math.min(hashPrefixes.length, other.hashPrefixes.length)];

        while (i < hashPrefixes.length && j < other.hashPrefixes.length) {
            if (hashPrefixes[i] < other.hashPrefixes[j]) {
                i++;
            } else if (hashPrefixes[i] > other.hashPrefixes[j]) {
                j++;
            } else {
                res[r++] = hashPrefixes[i++];
                j++;
            }
        }

        return new IndexHandle(id, Arrays.copyOf(res, r));
    }

    public IndexHandle union(IndexHandle other) {
        checkIds(other);

        int i = 0, j = 0, r = 0;
        var res = new int[hashPrefixes.length + other.hashPrefixes.length];

        while (i < hashPrefixes.length && j < other.hashPrefixes.length) {
            if (hashPrefixes[i] < other.hashPrefixes[j]) {
                res[r++] = hashPrefixes[i++];
            } else if (hashPrefixes[i] > other.hashPrefixes[j]) {
                res[r++] = other.hashPrefixes[j++];
            } else {
                res[r++] = hashPrefixes[i++];
                j++;
            }
        }

        while (i < hashPrefixes.length) {
            res[r++] = hashPrefixes[i++];
        }

        while (j < other.hashPrefixes.length) {
            res[r++] = other.hashPrefixes[j++];
        }

        return new IndexHandle(id, Arrays.copyOf(res, r));
    }

    private void checkIds(IndexHandle other) {
        if (!id.equals(other.id)) {
            throw new IllegalArgumentException("ids " + id + " and " + other.id + " differ");
        }
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        return o instanceof IndexHandle that && id.equals(that.id) && Arrays.equals(hashPrefixes, that.hashPrefixes);
    }

    @Override
    public int hashCode() {
        return 31 * id.hashCode() + Arrays.hashCode(hashPrefixes);
    }

    @Override
    public String toString() {
        return "IndexHandle{" +
                "id=" + id +
                ", hashPrefixes=[" + IntStream.of(hashPrefixes).mapToObj(h -> "0x" + Integer.toHexString(h)).collect(Collectors.joining(",")) +
                "]}";
    }
}
